src/analysis/bits-bounds-lattice.h

```diff
@@ -0,0 +1,144 @@
+
+    bool isBottom() const { return state == BOTTOM; }
+
+    void setUpperBound(Index upperBound) {
+      if (state != TOP) {
+        state = REGULAR_VALUE;
+        this->upperBound = upperBound;
+        constVal.reset();
+      }
+    }
+
+    void setUpperBound(Index upperBound, Literal constVal) {
+      if (state != TOP) {
+        state = REGULAR_VALUE;
+        this->upperBound = upperBound;
+        this->constVal = constVal;
+      }
+    }
+
+    void setTop() {
+      state = TOP;
+      upperBound = UINT32_MAX;
+      constVal.reset();
+    }
+
+    bool makeLeastUpperBound(const Element& other) {
+      if (other.isBottom() || isTop()) {
+        return false;
+      } else if (other.isTop()) {
+        setTop();
+        return true;
+      } else if (isBottom()) {
+        if (other.constVal.has_value()) {
+          setUpperBound(other.upperBound, other.constVal.value());
+        } else {
+          setUpperBound(other.upperBound);
+        }
+        return true;
+      }
+
+      if (upperBound < other.upperBound) {
+        if (other.constVal.has_value()) {
+          setUpperBound(other.upperBound, other.constVal.value());
+        } else {
+          setUpperBound(other.upperBound);
+        }
+        return true;
+      } else if (upperBound == other.upperBound && other.constVal.has_value()) {
+        if (constVal.has_value()) {
+          constVal.reset();
+        } else {
+          constVal = other.constVal.value();
+        }
+        return true;
+      }
+
+      return false;
+    }
+
+    void print(std::ostream& os) {
+      if (state == TOP) {
+        os << "TOP";
+      } else if (state == BOTTOM) {
+        os << "BOTTOM";
+      } else {
+        os << upperBound;
+        if (constVal.has_value()) {
+          os << " " << constVal.value();
+        }
+      }
+    }
+
+    friend MaxBitsLattice;
+  };
+
+  LatticeComparison compare(const Element& left, const Element& right) {
+    if (left.isTop()) {
+      if (right.isTop()) {
+        return LatticeComparison::EQUAL;
+      } else {
+        return LatticeComparison::GREATER;
+      }
+    } else if (right.isTop()) {
+      return LatticeComparison::LESS;
+    } else if (left.isBottom()) {
+      if (right.isBottom()) {
+        return LatticeComparison::EQUAL;
+      } else {
+        return LatticeComparison::LESS;
+      }
+    } else if (right.isBottom()) {
+      return LatticeComparison::GREATER;
+    }
+
+    if (left.upperBound > right.upperBound) {
+      return LatticeComparison::GREATER;
+    } else if (left.upperBound < right.upperBound) {
+      return LatticeComparison::LESS;
+    } else {
+      if (left.constVal.has_value()) {
+        if (right.constVal.has_value()) {
+          if (left.constVal.value() == right.constVal.value()) {
+            return LatticeComparison::EQUAL;
+          } else {
+            return LatticeComparison::NO_RELATION;
+          }
+        } else {
+          return LatticeComparison::LESS;
```

Comment:
If one side has a constant and the other does not, the correct thing to do is to determine the max bits of the constant and compare that to the max bits of the other side. `literal(8) < maxBits(3)`, but `literal(8) <> maxBits(2)` because we want `lub(literal(8), maxBits(2)) ==maxBits(3)`.

We've arrived at the part of the project where the Galois connections are actually relevant!